redisson/src/main/java/org/redisson/api/WorkerOptions.java

```diff
@@ -60,20 +60,21 @@ public WorkerOptions workers(int workers) {
         this.workers = workers;
         return this;
     }
-    
-    public BeanFactory getBeanFactory() {
-        return beanFactory;
+
+    public TaskInjector getTaskInjector() {
+        return injector;
     }
     
     /**
-     * Defines Spring BeanFactory instance to execute tasks with Spring's '@Autowired', 
+     * Defines injector to execute tasks with annotation e.g. Spring's '@Autowired',
      * '@Value' or JSR-330's '@Inject' annotation.
-     * 
-     * @param beanFactory - Spring BeanFactory instance
+     *
+     * @see org.redisson.spring.misc.BeanFactoryAdapter
+     * @param injector - a custom injector instance e.g. org.redisson.spring.misc.BeanFactoryAdapter
      * @return self instance
      */
-    public WorkerOptions beanFactory(BeanFactory beanFactory) {
-        this.beanFactory = beanFactory;
+    public WorkerOptions taskInjector(TaskInjector injector) {
```

Comment:
It's a breaking change for users who uses `beanFactory` option.

Reply:
@mrniko 

> It's a breaking change for users who uses beanFactory option. Why do we need to introduce taskInjector if quarkus-spring** libs added to classpath?

Indeed it is and I have no better solution for this problem at the moment. An alternative could be to put AutowiredAnnotationBeanPostProcessor and BeanFactory in the runtime project of the extension without any dependencies (x1) but this would be surprising to others and may lead to other problems with other extensions later on. If you have another idea feel free to share it.

If you don't remove AutowiredAnnotationBeanPostProcessor and BeanFactory you will see the following:

(in this stack trace) TasksRunnerService.java:318 --> `AutowiredAnnotationBeanPostProcessor bpp = new AutowiredAnnotationBeanPostProcessor();`

```
Detailed message:
Trace: 
	at parsing org.redisson.executor.TasksRunnerService.decode(TasksRunnerService.java:318)
Call path from entry point to org.redisson.executor.TasksRunnerService.decode(TaskParameters): 
	at org.redisson.executor.TasksRunnerService.decode(TasksRunnerService.java:282)
	at org.redisson.executor.TasksRunnerService.executeCallable(TasksRunnerService.java:213)
	at com.oracle.svm.reflect.RemoteExecutorService_executeCallable_381a82befffa37d303c23f37564ab3150e3aec4c_1376.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Method.java:566)
	at java.util.ServiceLoader$ProviderImpl.invokeFactoryMethod(ServiceLoader.java:737)
	at java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:720)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.initializeLazyValue(SystemPropertiesSupport.java:216)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.getProperty(SystemPropertiesSupport.java:169)
	at com.oracle.svm.core.jdk.Target_java_lang_System.getProperty(JavaLangSubstitutions.java:291)
	at com.oracle.svm.jni.JNIJavaCallWrappers.jniInvoke_VARARGS:Ljava_lang_System_2_0002egetProperty_00028Ljava_lang_String_2_00029Ljava_lang_String_2(generated:0)

com.oracle.svm.core.util.UserError$UserException: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved type during parsing: org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor. To diagnose the issue you can use the --allow-incomplete-classpath option. The missing type is then reported at run time when it is accessed the first time.
Detailed message:
Trace: 
	at parsing org.redisson.executor.TasksRunnerService.decode(TasksRunnerService.java:318)
Call path from entry point to org.redisson.executor.TasksRunnerService.decode(TaskParameters): 
	at org.redisson.executor.TasksRunnerService.decode(TasksRunnerService.java:282)
	at org.redisson.executor.TasksRunnerService.executeCallable(TasksRunnerService.java:213)
	at com.oracle.svm.reflect.RemoteExecutorService_executeCallable_381a82befffa37d303c23f37564ab3150e3aec4c_1376.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Method.java:566)
	at java.util.ServiceLoader$ProviderImpl.invokeFactoryMethod(ServiceLoader.java:737)
	at java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:720)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.initializeLazyValue(SystemPropertiesSupport.java:216)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.getProperty(SystemPropertiesSupport.java:169)
	at com.oracle.svm.core.jdk.Target_java_lang_System.getProperty(JavaLangSubstitutions.java:291)
	at com.oracle.svm.jni.JNIJavaCallWrappers.jniInvoke_VARARGS:Ljava_lang_System_2_0002egetProperty_00028Ljava_lang_String_2_00029Ljava_lang_String_2(generated:0)

	at com.oracle.svm.core.util.UserError.abort(UserError.java:82)
	at com.oracle.svm.hosted.FallbackFeature.reportAsFallback(FallbackFeature.java:233)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:798)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:582)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:495)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved type during parsing: org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor. To diagnose the issue you can use the --allow-incomplete-classpath option. The missing type is then reported at run time when it is accessed the first time.
Detailed message:
Trace: 
	at parsing org.redisson.executor.TasksRunnerService.decode(TasksRunnerService.java:318)
Call path from entry point to org.redisson.executor.TasksRunnerService.decode(TaskParameters): 
	at org.redisson.executor.TasksRunnerService.decode(TasksRunnerService.java:282)
	at org.redisson.executor.TasksRunnerService.executeCallable(TasksRunnerService.java:213)
	at com.oracle.svm.reflect.RemoteExecutorService_executeCallable_381a82befffa37d303c23f37564ab3150e3aec4c_1376.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Method.java:566)
	at java.util.ServiceLoader$ProviderImpl.invokeFactoryMethod(ServiceLoader.java:737)
	at java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:720)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.initializeLazyValue(SystemPropertiesSupport.java:216)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.getProperty(SystemPropertiesSupport.java:169)
	at com.oracle.svm.core.jdk.Target_java_lang_System.getProperty(JavaLangSubstitutions.java:291)
	at com.oracle.svm.jni.JNIJavaCallWrappers.jniInvoke_VARARGS:Ljava_lang_System_2_0002egetProperty_00028Ljava_lang_String_2_00029Ljava_lang_String_2(generated:0)

	at com.oracle.graal.pointsto.constraints.UnsupportedFeatures.report(UnsupportedFeatures.java:126)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:795)
	... 8 more
Caused by: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved type during parsing: org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor. To diagnose the issue you can use the --allow-incomplete-classpath option. The missing type is then reported at run time when it is accessed the first time.
	at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.reportUnresolvedElement(SharedGraphBuilderPhase.java:286)
	at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.handleUnresolvedType(SharedGraphBuilderPhase.java:242)
	at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.handleUnresolvedNewInstance(SharedGraphBuilderPhase.java:186)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genNewInstance(BytecodeParser.java:4643)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genNewInstance(BytecodeParser.java:4636)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBytecode(BytecodeParser.java:5451)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3451)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBlock(BytecodeParser.java:3258)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:1125)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:1019)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:84)
	at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.run(SharedGraphBuilderPhase.java:76)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:212)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
	at com.oracle.graal.pointsto.flow.AnalysisParsedGraph.parseBytecode(AnalysisParsedGraph.java:113)
	at com.oracle.svm.hosted.SVMHost.parseBytecode(SVMHost.java:647)
	at com.oracle.graal.pointsto.meta.AnalysisMethod.ensureGraphParsed(AnalysisMethod.java:592)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:163)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:304)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.createTypeFlow(MethodTypeFlow.java:313)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureTypeFlowCreated(MethodTypeFlow.java:302)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:103)
	at com.oracle.graal.pointsto.DefaultAnalysisPolicy$DefaultSpecialInvokeTypeFlow.onObservedUpdate(DefaultAnalysisPolicy.java:368)
	at com.oracle.graal.pointsto.flow.TypeFlow.notifyObservers(TypeFlow.java:470)
	at com.oracle.graal.pointsto.flow.TypeFlow.update(TypeFlow.java:542)
	at com.oracle.graal.pointsto.BigBang$2.run(BigBang.java:547)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:173)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	... 5 more
```

The spring dependencies org.springframework.beans.factory.BeanFactory is not part of the quarkus-spring-bean-api extension and I suspect that because of this AutowiredAnnotationBeanPostProcessor will not work during native compilation.

In my humble opinion it is not the best decision to have in a library like redisson any dependencies to a framework like spring because it may limit re-usability in non spring contexts. A better solution from my point of view to push all spring specifics to a redisson-spring project and keep the core library as minimal as possible.
In the context of building a "native-image"/executable it becomes a issue compilation.

(x1)
```
package org.springframework.beans.factory.annotation;
import org.springframework.beans.factory.BeanFactory;

public class AutowiredAnnotationBeanPostProcessor {

    public AutowiredAnnotationBeanPostProcessor() {

    }

    public void setBeanFactory(BeanFactory beanFactory) {
        
    }

    public void processInjection(Object bean) {
       
    }

}

package org.springframework.beans.factory;

public interface BeanFactory {
}


```

Reply:
What if create `TaskInjector` right in `public WorkerOptions beanFactory(BeanFactory beanFactory) {` method ?

Reply:
> What if create TaskInjector right in public WorkerOptions beanFactory(BeanFactory beanFactory) { method ?

let me try and see if it compiles.

Reply:
@mrniko see https://github.com/redisson/redisson/pull/3782/commits/fb5fa5800584bdf00e8f1612add16cee6ab671ae

Reply:
Sorry, I meant to remove `public WorkerOptions taskInjector(TaskInjector injector) {` and create  TaskInjector right in public WorkerOptions beanFactory(BeanFactory beanFactory) { method.

No deprecations and users won't be affected in this case.

Reply:
>Sorry, I meant to remove public WorkerOptions taskInjector(TaskInjector injector) { and create TaskInjector right in public WorkerOptions beanFactory(BeanFactory beanFactory) { method. No deprecations and users won't be affected in this case.



You mean? 

```
/**
     * Defines Spring BeanFactory instance to execute tasks with Spring's '@Autowired',
     * '@Value' or JSR-330's '@Inject' annotation.
     *
     * @param beanFactory - Spring BeanFactory instance
     * @return self instance
     */
    public WorkerOptions beanFactory(BeanFactory beanFactory) {
        this.injector = BeanFactoryAdapter.create(beanFactory);
        return this;
    }
```

and remove `public WorkerOptions taskInjector(TaskInjector injector)`

this would give you less options in supplying  alternative injector (non spring) in the future.
while what is there now keeps this option open.

So I am not sure why you would want that?

I can remove the deprecation notice if you don't intend to remove the spring dependency code from the Config and WorkerOptions class.

Reply:
deprecated annotation needs to be removed. Also comments describing taskInjector should be fixed. Please extract taskInjector to a separate PR.